src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,416 @@
+    if (value_count == 0) {
+      current_aggregate_value = std::optional<AggregateType>();
+    } else {
+      current_aggregate_value = *current_aggregate_value / value_count;
+    }
+  }
+  if constexpr (function == AggregateFunction::CountDistinct) {
+    current_aggregate_value = unique_values.size();
+  }
+  aggregate_null_values[aggregate_group_index] = !(current_aggregate_value.has_value());
+  if (current_aggregate_value.has_value()) {
+    aggregate_results[aggregate_group_index] = *current_aggregate_value;
+  }
+}
+
+std::shared_ptr<const Table> AggregateSort::_on_execute() {
+  const auto input_table = input_table_left();
+
+  // check for invalid aggregates
+  for (const auto& aggregate : _aggregates) {
+    if (!aggregate.column) {
+      if (aggregate.function != AggregateFunction::Count) {
+        Fail("Aggregate: Asterisk is only valid with COUNT");
+      }
+    } else {
+      DebugAssert(*aggregate.column < input_table->column_count(), "Aggregate column index out of bounds");
+      if (input_table->column_data_type(*aggregate.column) == DataType::String &&
+          (aggregate.function == AggregateFunction::Sum || aggregate.function == AggregateFunction::Avg)) {
+        Fail("Aggregate: Cannot calculate SUM or AVG on string column");
+      }
+    }
+  }
+
+  // collect groupy column definitions
+  for (const auto &column_id : _groupby_column_ids) {
+    _output_column_definitions.emplace_back(input_table->column_name(column_id),
+                                            input_table->column_data_type(column_id), true);
+  }
+
+  // collect aggregate column definitions
+  ColumnID column_index{0};
+  for (const auto& aggregate : _aggregates) {
+    const auto column = aggregate.column;
+
+    // Output column for COUNT(*). int is chosen arbitrarily.
+    const auto data_type = !column ? DataType::Int : input_table->column_data_type(*column);
+
+    resolve_data_type(
+        data_type, [&, column_index](auto type) { _write_aggregate_output(type, column_index, aggregate.function); });
+
+    ++column_index;
+  }
+
+  auto result_table = std::make_shared<Table>(_output_column_definitions, TableType::Data);
+
+  // handle empty input table
+  // if groupby columnds exist -> empty result
+  // else -> one row with default values
+  if (input_table->empty()) {
+    if (_groupby_column_ids.empty()) {
+      std::vector<AllTypeVariant> default_values;
+      for (const auto &aggregate : _aggregates) {
+        if (aggregate.function == AggregateFunction::Count || aggregate.function == AggregateFunction::CountDistinct) {
+          default_values.emplace_back(AllTypeVariant{0});
+        } else {
+          default_values.emplace_back(NULL_VALUE);
+        }
+      }
+      result_table->append(default_values);
+    }
+
+    return result_table;
+  }
+
+  // ToDo: Investigate if projection to used columns (aggregates + groupby) helps
+
+  // sort input table consecutively by the groupby columns (stable sort)
```

Comment:
> multiple sorting is not the fastest solution, but at least it works. sorting small ranges (for the second and all following groupby columns) should be faster imho

I completely agree. We chose this implementation because hyrise does not yet support multi-column sort, and writing an own specialised sort (operator) was out of (time) scope.
In the presentation we suggested some specialised sorting algorithms which should (in theory, at least) be much faster than the `std::sort` that is currently used by the Sort-Operator.
However, I'm not sure if this here is the best place for general sort optimization ideas?